stacker/plan2.py

```diff
@@ -0,0 +1,297 @@
+    @property
+    def completed(self):
+        """Returns True if the step is in a COMPLETE state."""
+        return self.status == COMPLETE
+
+    @property
+    def skipped(self):
+        """Returns True if the step is in a SKIPPED state."""
+        return self.status == SKIPPED
+
+    @property
+    def cancelled(self):
+        """Returns True if the step is in a CANCELLED state."""
+        return self.status == CANCELLED
+
+    @property
+    def failed(self):
+        """Returns True if the step is in a FAILED state."""
+        return self.status == FAILED
+
+    @property
+    def done(self):
+        """Returns True if the step is finished (either COMPLETE, SKIPPED or
+        CANCELLED, or FAILED)
+        """
+        return self.completed or self.skipped or self.cancelled or self.failed
+
+    @property
+    def ok(self):
+        """Returns True if the step is finished (either COMPLETE or SKIPPED)"""
+        return self.completed or self.skipped
+
+    @property
+    def submitted(self):
+        """Returns True if the step is SUBMITTED, COMPLETE, or SKIPPED."""
+        return self.status >= SUBMITTED
+
+    def set_status(self, status):
+        """Sets the current step's status.
+        Args:
+            status (:class:`Status <Status>` object): The status to set the
+                step to.
+        """
+        if status is not self.status:
+            logger.debug("Setting %s state to %s.", self.stack.name,
+                         status.name)
+            logger.info("%s %s", status.name, self.short_name)
+            self.status = status
+            self.last_updated = time.time()
+
+    def complete(self):
+        """A shortcut for set_status(COMPLETE)"""
+        self.set_status(COMPLETE)
+
+    def skip(self):
+        """A shortcut for set_status(SKIPPED)"""
+        self.set_status(SKIPPED)
+
+    def submit(self):
+        """A shortcut for set_status(SUBMITTED)"""
+        self.set_status(SUBMITTED)
+
+
+def build_plan(description=None, steps=None,
+               step_names=None, reverse=False):
+    graph = build_graph(steps)
+
+    # If we want to execute the plan in reverse (e.g. Destroy), transpose the
+    # graph.
+    if reverse:
+        graph = graph.transposed()
+
+    # If we only want to build a specific target, filter the graph.
+    if step_names:
+        nodes = []
+        for step_name in step_names:
+            for step in steps:
+                if step.short_name == step_name:
+                    nodes.append(step.name)
+        graph = graph.filtered(nodes)
+
+    return Plan(description=description, graph=graph)
+
+
+def build_graph(steps):
+    """Builds a graph of steps.
+    Args:
+        steps (list): a list of :class:`Step` objects to execute.
+    """
+
+    graph = Graph()
+
+    for step in steps:
+        graph.add_step(step)
+
+    for step in steps:
+        for dep in step.requires:
+            graph.connect(step, dep)
+
+    return graph
+
+
+class Graph(object):
```

Comment:
Curious if this should just be combined with the `Plan` class?

Reply:
I originally started with that, but I think it's cleaner to keep these things separate:

1. `Graph` encapsulates a graph of `Step` objects, and has operations to make working on the graph itself easy (transposition, walking, etc).
2. `Plan` is more like a convenience class for Actions to consume (dump, outline, etc).

`Plan` could probably eventually get removed, as most of the methods on it may be better somewhere else, but I primarily kept it around to keep API compatibility with the existing `Plan` implementation.